### ISSUE TITLE
Tracing not working in Blazor WebAssembly application

### ISSUE THREAD

**Issue by mkuennek** [open-telemetry/opentelemetry-dotnet#2816]
# Bug Report

OpenTelemetry.Exporter.Console 1.2.0-rc1

Runtime version (e.g. `net461`, `net48`, `netcoreapp3.1`, `net5.0` etc. You can
find this information from the `*.csproj` file):

net6.0 Blazor app running as WebAssembly

## Symptom

Adding tracing to a Blazor wasm app results in System.PlatformNotSupportedException during runtime as Process.GetCurrentProcess() is not supported in Blazor wasm. The stacktrace is:

System.AggregateException: One or more errors occurred. (The type initializer for 'OpenTelemetry.Resources.ResourceBuilder' threw an exception.) ---> System.TypeInitializationException: The type initializer for 'OpenTelemetry.Resources.ResourceBuilder' threw an exception. ---> System.PlatformNotSupportedException: System.Diagnostics.Process is not supported on this platform. at System.Diagnostics.Process.GetCurrentProcess() at OpenTelemetry.Resources.ResourceBuilder..cctor() --- End of inner exception stack trace --- at OpenTelemetry.Trace.TracerProviderBuilderBase..ctor() at OpenTelemetry.Trace.TracerProviderBuilderSdk..ctor() at OpenTelemetry.Sdk.CreateTracerProviderBuilder() at Program.
$(String[] args) in C:\Sources\Privat\OtelBlazorTracingBug\Program.cs:line 12 --- End of inner exception stack trace ---

The reasons seems to be the in the DefaultResource property of the ResourceBuilder class, see https://github.com/open-telemetry/opentelemetry-dotnet/blob/d56f925a734aacfb84cca14a8992655e774867f1/src/OpenTelemetry/Resources/ResourceBuilder.cs#L33

What did you expect to see?

Tracing to work in a Blazor wasm app.

What did you see instead?

Runtime exception

## Reproduce

See the following repo for a simple reproduction:
https://github.com/mkuennek/OtelBlazorTracingBug

## Additional Context

Also discovered elsewhere, see https://wolfgang-ziegler.com/blog/otel-dotnet-blazor

**Comment by ProgrammerAl** [comment]
Adding a note to this issue so others don't have to research like I just did. It's possible this issue is now fixed after #3405 was merged in. I haven't been able to test it yet because the current version of code hasn't been pushed to a new NuGet package. #3405 was merged on June 24, 2022, and the latest NuGet was published June 3rd, 2022. So once the next NuGet is published we can test again on Blazor WebAssembly.

**Comment by JoeShook** [comment]
While @ProgrammerAl pointed out the default service name issue may be fixed after #3405 the lack of threading support in Blazor WASM would still be a blocker.  But...  It appears further experimentation with my [sample project](https://github.com/JoeShook/opentelemetry-dotnet/tree/fixup_blazor_webassembly) could be revisited now that there is experimental support for threading in Blazor WASM.  [wasm-experimental: Experimental support for WebAssembly multithreading](https://github.com/dotnet/core/issues/7717)

**Comment by snow-jallen** [comment]
Is there a way to do this now?

**Comment by vincentnl** [comment]
Here is someone worth of buying a coffee - hope somebody can pick this up.
https://wolfgang-ziegler.com/blog/otel-dotnet-blazor

**Comment by vincentnl** [comment]
I am working on a proof of concept to get 'poor man's' tracing up and running using the opentelemetry endpoint as a target.
Got it working - and basic tracing enabled.

Feel free to comment or add:

https://github.com/vincentnl/TracesForBlazor

**Comment by gunawanyuwono** [comment]
> I am working on a proof of concept to get 'poor man's' tracing up and running using the opentelemetry endpoint as a target. Got it working - and basic tracing enabled.
> 
> Feel free to comment or add:
> 
> https://github.com/vincentnl/TracesForBlazor

Hi vincentnl, can we use this library for cross-origin (other host project for api target of web assembly)?
So we have 2 project
1.Project webassembly
2.Project Web API
web assembly client will send request to Web API Project (I use ASP.NET Web API Core 6.0)


Regards,

Gunawan

**Comment by vincentnl** [comment]
Dear Gunawan,

I think you are referring to distributed tracing? As that the trace spans different parts of the system.

That is not implemented - as stated it is purely Poor man’s standin for blazor wasm.

It will not work over server or hybrid.

On Jan 3, 2024, at 05:17, gunawanyuwono ***@***.***> wrote:

﻿

I am working on a proof of concept to get 'poor man's' tracing up and running using the opentelemetry endpoint as a target. Got it working - and basic tracing enabled.

Feel free to comment or add:

https://github.com/vincentnl/TracesForBlazor

Hi vincentnl, can we use this library for cross-origin (other host project for api target of web assembly)?
So we have 2 project
1.Project webassembly
2.Project Web API
web assembly client will send request to Web API Project (I use ASP.NET Web API Core 6.0)

Regards,

Gunawan

**Comment by luisabreu** [comment]
Hello guys,

any news on this issue? I mean, has there been any developments on the  blazor wasm /distributed tracing incompatibilities lately?

**Comment by madskonradsen** [comment]
Multithreading has now been postponed to .Net10 - https://github.com/dotnet/aspnetcore/issues/17730#issuecomment-2059602250 - and it could be postponed even further(was originally planned for .Net5).

**Comment by verdie-g** [comment]
@cijothomas could https://github.com/open-telemetry/opentelemetry-dotnet/pull/5838 also solve this issue as we are removing custom threads from OTEL?

**Comment by REscobar** [comment]
I can confirm that a Blazor WASM app is able to push traces to a Zipkin endpoint using the changes proposed in this pr #5838 and the exporter changes proposed in this branch https://github.com/verdie-g/opentelemetry-dotnet/commit/02eb4503bcd5dfcb415351f1791ddc3b913eb5a7, only the batch exporter works though, since simple exporter is still sync over async on that branch, I did not test other signals or exporters.

Some screenshots:
![image](https://github.com/user-attachments/assets/fde1de5a-9e2d-4977-869c-c7a52728cf77)
![image](https://github.com/user-attachments/assets/5b57590f-6af8-48a4-912a-66113691e3e9)
![image](https://github.com/user-attachments/assets/9ae55fe8-ad4e-442a-bb8f-6440849cd79c)

Configuration on blazor app

```csharp
builder.Services.AddOpenTelemetry()
    .ConfigureResource(r =>
    {
        r.AddService("BlazorWASMFrontend");
    })
    .WithTracing(tracerProviderBuilder =>
        tracerProviderBuilder.AddHttpClientInstrumentation()
        .AddSource("BlazorWASMFrontend")
        //.AddConsoleExporter()
        .AddZipkinExporter(o =>
        {
            o.ExportProcessorType = ExportProcessorType.Batch;
            o.HttpClientFactory = () =>
            {
                return new HttpClient();
            };
            o.Endpoint = new Uri("http://localhost:9411/api/v2/spans");
        })
    )
```

**Comment by jeromelaban** [comment]
I'm taking another stab at this issue with https://github.com/open-telemetry/opentelemetry-dotnet/pull/6379 in order to support WebAssembly specifically (to support for blazor, Uno Platform and others), while still keeping the thread support code available to reduce possible regressions.

**Comment by mvanhilst-msft** [comment]
@jeromelaban I used your changes with the usual OpenTelemetry Program.cs setup ([here](https://github.com/mvanhilst-msft/OrleansExperiment/blob/main/OrleansExperiment.Frontend/Program.cs)) for my Blazor WASM standalone app using `.AddHttpClientInstrumentation()` to test out tracing. It worked great! ... but only after I forced the DI container to create the `TracerProvider` and `MeterProvider` after the call to `builder.Build()`:

```cs
var host = builder.Build();
host.Services.GetRequiredService<MeterProvider>();
host.Services.GetRequiredService<TracerProvider>();
await host.RunAsync();
```

Is this something that the OpenTelemetry ASP.NET Core integration usually takes care of? Should I open a separate issue for handling this automatically in Blazor WASM?

**Comment by alexaka1** [comment]
@mvanhilst-msft Yes, afaik this is an asp.net core thing, so in other apps (console, wasm, etc), you have to manually instantiate the trace and meter providers. (Not the logging for whatever reason, that works regardless)

**Comment by mvanhilst-msft** [comment]
> [@mvanhilst-msft](https://github.com/mvanhilst-msft) Yes, afaik this is an asp.net core thing, so in other apps (console, wasm, etc), you have to manually instantiate the trace and meter providers. (Not the logging for whatever reason, that works regardless)

I dug a bit deeper, because I know it works fine in console apps that use `HostApplicationBuilder`. The limitation is that `WebAssemblyHostBuilder` ignores `IHostedService`s, so the `TelemetryHostedService` is never initialized. In fact, it looks like there are no lifecycle hooks at all, so I don't see an alternative way for OpenTelemetry to self-initialize via DI.

**Comment by oroztocil** [comment]
@mvanhilst-msft Hi, we created https://github.com/dotnet/aspnetcore/issues/63637 to track the request for supporting hosted services in `WebAssemblyHost` apps.
